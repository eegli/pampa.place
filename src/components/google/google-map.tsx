import {MapService} from '@/services/google';
import Box from '@mui/material/Box';
import {memo, ReactNode, useEffect, useRef} from 'react';

type BoundsProps = {
  bounds: {
    SW: google.maps.LatLngLiteral;
    NE: google.maps.LatLngLiteral;
  };
  padding?: number;
};

type CenterProps = {
  center: google.maps.LatLngLiteral;
  zoom?: number;
};

export type GoogleMapProps = {
  id?: string;
  children?: ReactNode;
  onMount?: (map: google.maps.Map) => unknown;
  onUnmount?: (map: google.maps.Map) => unknown;
} & (BoundsProps | CenterProps);

const _GoogleMap = ({
  children,
  id = 'google-map',
  onMount,
  onUnmount,
  ...rest
}: GoogleMapProps) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      const unmountMap = MapService.mount(ref.current);
      console.info('%cGOOGLE MAP MOUNT', 'color: yellow');

      if ('bounds' in rest) {
        /* Order in constructor is important! SW, NE  */
        const bounds = new google.maps.LatLngBounds(
          rest.bounds.SW,
          rest.bounds.NE
        );
        google.maps.event.addListenerOnce(MapService.map, 'idle', () => {
          MapService.map.fitBounds(bounds, rest.padding || 0);
        });
      } else {
        MapService.map.setCenter(rest.center);
        MapService.map.setZoom(rest.zoom || 0);
      }

      if (typeof onMount === 'function') {
        onMount(MapService.map);
      }

      return () => {
        // Any cleanup logic
        if (typeof onUnmount === 'function') {
          onUnmount(MapService.map);
        }

        unmountMap();
        console.info('%cGOOGLE MAP UNMOUNT', 'color: yellow');
      };
    }
  }, [ref, onMount, onUnmount, rest]);

  return (
    <Box id={id} data-testid={id} ref={ref} height="100%" width="100%">
      {children}
    </Box>
  );
};

export const GoogleMap = memo(_GoogleMap, (prev, next) => prev.id === next.id);
