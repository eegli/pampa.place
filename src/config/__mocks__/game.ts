import {GameConfig} from '../types';

export const config: GameConfig = {
  maxPlayers: 4,
  rounds: [1, 2, 3],
  roundsDefault: 2,
  timeLimits: [30, 60, 120, -1],
  timeLimitsDefault: 30,
};
