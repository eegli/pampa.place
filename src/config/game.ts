type GameConfig = {
  maxPlayers: number;
  rounds: [number, number, number];
  roundsDefault: number;
  timeLimits: [number, number, number, number];
  timeLimitsDefault: number;
};

export const config: GameConfig = {
  maxPlayers: 4,
  rounds: [1, 3, 5],
  roundsDefault: 3,
  // -1 for unlimited time
  timeLimits: [30, 60, 120, -1],
  timeLimitsDefault: 60,
};
