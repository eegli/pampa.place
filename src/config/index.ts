import { defaultSettings, DefaultSettings } from './defaults';
import customMaps from './maps';
import { markerColors } from './markers';

export type MarkerColors = Record<string, string>;

export type MapBounds = Record<
  'latMax' | 'latMin' | 'lngMax' | 'lngMin',
  number
>;
export type MapLatLng = Record<'lat' | 'lng', number>;
export type MapCenter = { center: MapLatLng };
export type MapEdges = Record<'NE' | 'SE' | 'SW' | 'NW', MapLatLng>;

export type UserMapConfig = MapBounds & { __name: string };

export type MapData = UserMapConfig & MapCenter & MapEdges;

interface Config {
  maps: MapData[];
  defaults: DefaultSettings;
  markers: string[];
}

const makeConfig = (): Config => {
  return {
    markers: Object.values(markerColors),
    maps: customMaps.reduce((acc, curr) => {
      acc.push({
        ...curr,
        center: {
          lat: (curr.latMax + curr.latMin) / 2,
          lng: (curr.latMax + curr.latMin) / 2,
        },
        NE: { lat: curr.latMax, lng: curr.lngMax },
        SE: { lat: curr.latMin, lng: curr.lngMax },
        SW: { lat: curr.latMin, lng: curr.lngMin },
        NW: { lat: curr.latMax, lng: curr.lngMin },
      });
      return acc;
    }, [] as MapData[]),
    defaults: defaultSettings,
  };
};

export default makeConfig();
