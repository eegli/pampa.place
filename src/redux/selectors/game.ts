import { createSelector } from '@reduxjs/toolkit';
import { RootState } from '../store';

export const getActivePlayer = (s: RootState) => s.game.players[0];

export const getPlayerCount = (s: RootState) => s.game.players.length;

// TODO use memoization once redux toolkit exports reselect 4.1
export const getRoundScores = createSelector(
  [
    (s: RootState) => s.game.scores,
    (s: RootState) => s.game.rounds.current - 1,
  ],
  (scores, currentRound) => {
    /* Sort on a copy, otherwise, Redux isn't happy: https://stackoverflow.com/a/66721870 */
    return [...scores[currentRound]].sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

export const getTotalScores = createSelector(
  [(s: RootState) => s.game.scores, (s: RootState) => s.game.players],
  (scores, players) => {
    const final = Array.from({ length: players.length }, (_, idx) => ({
      name: players[idx],
      score: 0,
    }));

    scores.forEach(round => {
      round.forEach((playerResult, playerIdx) => {
        final[playerIdx].score += playerResult.score;
      });
    });

    return final.sort((a, b) =>
      a.score > b.score ? -1 : b.score > a.score ? 1 : 0
    );
  }
);

const s = [
  [
    { name: 'e', round: 1, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 1, selected: 'x', dist: 1, score: 999 },
  ],
  [
    { name: 'e', round: 2, selected: 'x', dist: 1, score: 999 },
    { name: 'x', round: 2, selected: 'x', dist: 1, score: 999 },
  ],
];

export const shouldRequestNewSV = (s: RootState) =>
  s.game.rounds.progress === 0;

export const isFinished = (s: RootState) =>
  s.game.rounds.progress === s.game.players.length &&
  s.game.rounds.current === s.game.rounds.total;
