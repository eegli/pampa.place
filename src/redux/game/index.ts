import {config} from '@/config/game';
import {MAPS, MAP_IDS} from '@/config/maps';
import {LatLngLiteral} from '@/config/types';
import {calcDist, calcScore} from '@/utils/geo';
import {OrNull} from '@/utils/types';
import {createSlice, PayloadAction} from '@reduxjs/toolkit';

export type Result = {
  name: string;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

export enum STATUS {
  UNINITIALIZED = 'UNINITIALIZED',
  ROUND_STARTED = 'ROUND_STARTED',
  PENDING_PLAYER = 'PENDING_PLAYER',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: string[];
  // Store the results for each round in the outer array. The innner
  // array consists of player result objects.
  scores: Result[][];
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

const initialState: GameState = {
  status: STATUS.UNINITIALIZED,
  mapId: MAP_IDS[0].id,
  mapName: MAP_IDS[0].name,
  players: [],
  scores: [],
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players = action.payload.filter(Boolean).map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].feature.properties.name;
    },
    initGame(state) {
      if (!state.players.length) {
        state.players = ['Player 1'];
      }
      state.scores = [];
      // Empty score array for first round
      state.scores.push([]);
      state.rounds.current = 1;
      state.rounds.progress = 0;
      state.status = STATUS.PENDING_PLAYER;
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].area, dist);
        selected = payload.selected;
      }

      state.scores[state.rounds.current - 1].push({
        name: state.players[0],
        selected,
        dist,
        score,
      });

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.pop();
      if (last) state.players.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.length) {
        // Display popup
        state.status = STATUS.PENDING_PLAYER;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
    resetRound(state) {
      // Reset score for this round
      state.scores[state.rounds.current - 1] = [];

      // Reset round progress
      state.rounds.progress = 0;

      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.pop();
        if (playerName) {
          state.players.unshift(playerName);
        }
      }

      // Set zero for current round scores
      state.status = STATUS.PENDING_PLAYER;
    },
    endRound(state) {
      // Game has ended
      if (state.rounds.total === state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        // Prepare for next round
        state.rounds.current++;
        state.rounds.progress = 0;
        state.scores.push([]);
        state.status = STATUS.PENDING_PLAYER;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  endRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice;
