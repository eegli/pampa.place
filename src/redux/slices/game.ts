import { config } from '@/config/game';
import { CUSTOM_MAP_IDS, LatLngLiteral, MAPS } from '@/config/maps';
import { calcDist, calcScore } from '@/utils/geo';
import { OrNull } from '@/utils/types';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export type Result = {
  round: number;
  selected: OrNull<LatLngLiteral>;
  dist: number;
  score: number;
};

type Player = {
  totalScore: number;
} & { results: Result[] };

export enum STATUS {
  PENDING_START = 'PENDING_START', // Uninitialized
  ROUND_STARTED = 'ROUND_STARTED',
  INTERMISSION = 'INTERMISSION',
  ROUND_ENDED = 'ROUND_ENDED',
  FINISHED = 'FINISHED',
}

export interface GameState {
  status: STATUS;
  mapId: string;
  mapName: string;
  players: {
    names: string[];
    scores: Record<string, Player>;
  };
  timeLimit: number;
  rounds: {
    current: number;
    progress: number;
    total: number;
  };
}

export const initialState: GameState = {
  status: STATUS.PENDING_START,
  mapId: CUSTOM_MAP_IDS[0].id,
  mapName: CUSTOM_MAP_IDS[0].name,
  players: {
    names: [],
    scores: {},
  },
  timeLimit: config.timeLimitsDefault,
  rounds: {
    current: 1,
    total: config.roundsDefault,
    progress: 0,
  },
};

export const gameSlice = createSlice({
  name: 'game',
  initialState,

  // The active player is always at index 0, and players are rotated
  // when they are done
  reducers: {
    reset(state) {
      state = initialState;
    },
    setPlayers(state, action: PayloadAction<string[]>) {
      state.players.names = action.payload
        .filter(Boolean)
        .map(el => el.slice(0, 24));
    },
    setRounds(state, action: PayloadAction<number>) {
      state.rounds.total = action.payload;
    },
    setTimeLimit(state, action: PayloadAction<number>) {
      state.timeLimit = action.payload;
    },
    setMap(state, action: PayloadAction<string>) {
      state.mapId = action.payload;
      state.mapName = MAPS[action.payload].geo.properties.name;
    },
    initGame(state) {
      if (!state.players.names.length) {
        state.players.names = ['Player 1'];
      }

      state.players.scores = {};

      state.players.names.forEach(name => {
        state.players.scores[name] = { totalScore: 0, results: [] };
      });

      state.rounds.current = 1;
      state.rounds.progress = 0;

      state.status = STATUS.INTERMISSION;
    },
    finishRound(state) {
      state.rounds.current++;
      state.rounds.progress = 0;

      // Game has ended
      if (state.rounds.total < state.rounds.current) {
        state.status = STATUS.FINISHED;
      } else {
        state.status = STATUS.INTERMISSION;
      }
    },
    startRound(state) {
      state.status = STATUS.ROUND_STARTED;
    },
    resetRound(state) {
      // Rotate to first player again
      for (let i = 0; i < state.rounds.progress; i++) {
        const playerName = state.players.names.pop();
        if (playerName) {
          // Remove score from this round
          const result = state.players.scores[playerName].results.pop();

          // Subtract score from this round from total score
          if (result)
            state.players.scores[playerName].totalScore -= result.score;

          // Rotate player
          state.players.names.unshift(playerName);
        }
      }

      // Reset round progress
      state.rounds.progress = 0;

      // Set zero for current round scores
      state.status = STATUS.INTERMISSION;
    },
    setPlayerScore(
      state,
      {
        payload,
      }: PayloadAction<{
        selected: OrNull<LatLngLiteral>;
        initial: OrNull<LatLngLiteral>;
      }>
    ) {
      // Set current round score
      const player = state.players.scores[state.players.names[0]];

      // Payload when the user fails to set a location in time
      let score = 0;
      let dist = -1;
      let selected: OrNull<LatLngLiteral> = null;

      // User managed to set a location
      if (payload.selected && payload.initial) {
        dist = calcDist(payload.initial, payload.selected);
        score = calcScore(MAPS[state.mapId].computed.area, dist);
        selected = payload.selected;
      }

      const newScore = player.totalScore + score;
      const existingResults = player.results;
      state.players.scores[state.players.names[0]] = {
        totalScore: newScore,
        results: [
          ...existingResults,
          {
            round: state.rounds.current,
            selected,
            dist,
            score,
          },
        ],
      };

      // Set round progress
      state.rounds.progress++;

      // Rotate players
      const last = state.players.names.pop();
      if (last) state.players.names.unshift(last);

      // Same round, wait for player change
      if (state.rounds.progress < state.players.names.length) {
        // Display popup
        state.status = STATUS.INTERMISSION;
        // Round is over, reset round progress
      } else {
        // Update overall score and force a new random location
        state.status = STATUS.ROUND_ENDED;
      }
    },
  },
});

export const {
  reset,
  initGame,
  startRound,
  finishRound,
  resetRound,
  setPlayerScore,
  setRounds,
  setMap,
  setPlayers,
  setTimeLimit,
} = gameSlice.actions;
export default gameSlice.reducer;
