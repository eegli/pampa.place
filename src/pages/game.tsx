import { NextPage } from 'next';
import { useRouter } from 'next/dist/client/router';
import { useEffect } from 'react';
import Error from '../components/error';
import Play from '../components/play/play';
import RoundEnd from '../components/round/round.end';
import RoundIntermission from '../components/round/round.intermission';
import RoundResult from '../components/round/round.result';
import { STATUS } from '../redux/game';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { getRandomStreetView } from '../redux/position';
import { PageContentContainer } from '../styles';

const Game: NextPage = () => {
  const dispatch = useAppDispatch();
  const router = useRouter();

  const initialized = useAppSelector(({ game }) => game.initialized);
  const status = useAppSelector(({ game }) => game.status);
  const positionError = useAppSelector(({ position }) => position.error);

  useEffect(() => {
    if (!initialized) {
      router.push('/');
    }
  }, [initialized]);

  async function handleRetry() {
    await dispatch(getRandomStreetView({ radius: 1000 }));
  }

  function render() {
    switch (status) {
      case STATUS.INTERMISSION:
        return <RoundIntermission />;
      case STATUS.ROUND_STARTED:
        return <Play />;
      case STATUS.ROUND_ENDED:
        return <RoundEnd />;
      case STATUS.FINISHED:
        return <RoundResult />;

      default:
        return <div />;
    }
  }

  // If there is a header included, the page content container should
  // take the remaining height. The header is 48p by default
  /* 
   <GameHeader />
      <PageContentContainer height="calc(100% - 48px)">
        {render()}
      </PageContentContainer>
    </>
  
  */
  return (
    <>
      <PageContentContainer height="100%">
        {positionError ? (
          <Error
            callback={handleRetry}
            title="Error getting Street View data"
            info="This is likely because the map you chose is a little
                too small or has little Street View coverage."
            reason={positionError.message}
          />
        ) : (
          render()
        )}
      </PageContentContainer>
    </>
  );
};

export default Game;
