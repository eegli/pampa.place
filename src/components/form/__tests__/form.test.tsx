import {config} from '@/config/game';
import {STATUS} from '@/redux/game/game.slice';
import {
  ByRoleOptions,
  createMockStore,
  fireEvent,
  render,
  screen,
} from '@/tests/test-utils';
import React from 'react';
import {Gmap} from '../../../services/google-map';
import Form from '../form';
import FormMapSelect from '../form.map-select';
import FormPlayers from '../form.players';
import FormRoundSelect from '../form.round-select';
import FormTimeLimitSelect from '../form.time-select';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      push: jest.fn(),
    };
  },
}));

function getRadios(opts?: ByRoleOptions) {
  return screen.getAllByRole('radio', opts);
}

describe('Form', () => {
  function querySubmitButton() {
    return screen.getByRole('button', {name: /start/i});
  }

  it('has submit button', () => {
    render(<Form />);
    const submit = querySubmitButton();
    expect(submit).toBeInTheDocument();
  });

  it('inits game on click', () => {
    const store = createMockStore();
    render(<Form />, store);
    const submit = querySubmitButton();
    fireEvent.click(submit);

    expect(store.getState().game.status).not.toEqual(STATUS.UNINITIALIZED);
  });
});

describe('Form, player name input', () => {
  function queryPlayerInput() {
    return screen.queryAllByLabelText(/^player/i);
  }

  it('always renders at least one player input field', () => {
    render(<FormPlayers />);
    expect(queryPlayerInput()).toHaveLength(1);
    expect(queryPlayerInput()[0]).toHaveValue('');
  });
  it('does not create more inputs than configured', () => {
    render(<FormPlayers />);

    for (let i = 0; i < config.maxPlayers + 2; i++) {
      const inputs = queryPlayerInput();
      if (inputs[i]) {
        fireEvent.change(inputs[i], {target: {value: `player ${i}`}});
      } else {
        break;
      }
    }
    expect(queryPlayerInput()).toHaveLength(config.maxPlayers);
  });
  it('truncates long and filters invalid player names', () => {
    render(<FormPlayers />);
    fireEvent.change(queryPlayerInput()[0], {
      target: {value: `eric eric eric eric eric eric eric eric eric eric`},
    });
    expect(queryPlayerInput()[0]).toHaveValue('eric eric eric eric eric');
    fireEvent.change(queryPlayerInput()[1], {
      target: {value: 'player 2'},
    });
    fireEvent.change(queryPlayerInput()[0], {
      target: {value: ''},
    });
    expect(queryPlayerInput()[0]).toHaveValue('player 2');
  });
});

describe('Form, round select', () => {
  it('renders round select options', () => {
    render(<FormRoundSelect />);
    expect(getRadios({checked: true})).toHaveLength(1);
    expect(getRadios()).toHaveLength(config.rounds.length);
  });

  it('updates round select radio buttons', () => {
    render(<FormRoundSelect />);
    const roundInputs = getRadios();
    fireEvent.click(roundInputs[0]);
    expect(getRadios()[0]).toHaveProperty('checked', true);
    fireEvent.click(roundInputs[1]);
    expect(getRadios()[1]).toHaveProperty('checked', true);
  });
});

describe('Form, duration select', () => {
  it('renders duration select options', () => {
    const {container} = render(<FormTimeLimitSelect />);
    expect(container).toMatchSnapshot();

    expect(getRadios({checked: true})).toHaveLength(1);
    expect(getRadios()).toHaveLength(config.timeLimits.length);
  });

  it('updates duration select radio buttons', () => {
    render(<FormTimeLimitSelect />);
    const durationInputs = getRadios();
    fireEvent.click(durationInputs[0]);
    expect(getRadios()[0]).toHaveProperty('checked', true);
    fireEvent.click(durationInputs[1]);
    expect(getRadios()[1]).toHaveProperty('checked', true);
  });
});

// TODO
describe('Form, map selection and preview', () => {
  const toggleSpy = jest.spyOn(Gmap, 'toggle');

  it('displays maps', () => {
    render(<FormMapSelect />);
    const mapSelection = screen.getByRole('button', {name: /custom map/i});
    expect(mapSelection).toBeInTheDocument();
  });
  it('displays map preview', () => {
    render(<FormMapSelect />);
    const previewMapButton = screen.getByTestId('map-preview-button');
    fireEvent.click(previewMapButton);
    expect(toggleSpy).toHaveBeenCalledTimes(1);
  });
});
