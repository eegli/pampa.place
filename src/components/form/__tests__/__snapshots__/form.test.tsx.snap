// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form, duration select renders duration select options 1`] = `
<div>
  <div
    id="__GMAP__CONTAINER__"
  >
    <div
      id="__GMAP__"
      style="height: 100%;"
    />
  </div>
  <div
    id="__GSTV__CONTAINER__"
  >
    <div
      id="__GSTV__"
      style="height: 100%;"
    />
  </div>
  <fieldset
    class="MuiFormControl-root css-18cl080-MuiFormControl-root"
  >
    <legend
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-vku667-MuiFormLabel-root"
    >
      Time limit
    </legend>
    <div
      class="MuiFormGroup-root MuiFormGroup-row css-qfz70r-MuiFormGroup-root"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-jzcd7z-MuiFormControlLabel-root"
      >
        <span
          class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-1yq3e2k-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="form-time-select"
            type="radio"
            value="30"
          />
          <span
            class="css-hyxlzm"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          30s
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-jzcd7z-MuiFormControlLabel-root"
      >
        <span
          class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root Mui-checked css-1yq3e2k-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="form-time-select"
            type="radio"
            value="60"
          />
          <span
            class="css-hyxlzm"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          1m
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-jzcd7z-MuiFormControlLabel-root"
      >
        <span
          class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-1yq3e2k-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="form-time-select"
            type="radio"
            value="120"
          />
          <span
            class="css-hyxlzm"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          2m
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-jzcd7z-MuiFormControlLabel-root"
      >
        <span
          class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-1yq3e2k-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            name="form-time-select"
            type="radio"
            value="-1"
          />
          <span
            class="css-hyxlzm"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          unlimited
        </span>
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`Form, map preview renders map options 1`] = `
<div
  aria-hidden="true"
>
  <div
    id="__GMAP__CONTAINER__"
  >
    <div
      id="__GMAP__"
      style="height: 100%;"
    />
  </div>
  <div
    id="__GSTV__CONTAINER__"
  >
    <div
      id="__GSTV__"
      style="height: 100%;"
    />
  </div>
</div>
`;

exports[`Form, map preview renders map options 2`] = `
<div>
  <div
    id="__GMAP__CONTAINER__"
  >
    <div
      id="__GMAP__"
      style="height: 100%;"
    />
  </div>
  <div
    id="__GSTV__CONTAINER__"
  >
    <div
      id="__GSTV__"
      style="height: 100%;"
    />
  </div>
</div>
`;

exports[`Form, map select renders map options 1`] = `
Array [
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="form-map-select"
    class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1coxseh-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
    id="form-map-select"
    role="button"
    tabindex="0"
  >
    Alpen
  </div>,
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1hf04jj-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      aria-label="Preview map"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-mui-internal-clone-element="true"
      data-testid="VisibilityIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>,
]
`;
